storage/SMAIndexSubBlock.cpp

```diff
@@ -0,0 +1,670 @@
+  // For each entry, clear the min and max typed values (frees out of line data).
+  for (std::size_t indexed_attribute_num = 0;
+       indexed_attribute_num < indexed_attributes_;
+       ++indexed_attribute_num) {
+    SMAEntry &entry = entries_[indexed_attribute_num];
+    if (!TypedValue::RepresentedInline(entry.type_id)) {
+      if (entry.min_entry_ref.valid) {
+        entry.min_entry_ref.value.clear();
+      }
+      if (entry.max_entry_ref.valid) {
+        entry.max_entry_ref.value.clear();
+      }
+    }
+  }
+}
+
+bool SMAIndexSubBlock::DescriptionIsValid(const CatalogRelationSchema &relation,
+                                          const IndexSubBlockDescription &description) {
+  if (!description.IsInitialized()) {
+    return false;
+  }
+  if (description.sub_block_type() != IndexSubBlockDescription::SMA) {
+    return false;
+  }
+  // Must have at least one indexed attribute.
+  if (description.ExtensionSize(SMAIndexSubBlockDescription::indexed_attribute_id) == 0) {
+    return false;
+  }
+  // Check that all key attributes exist.
+  for (int indexed_attribute_num = 0;
+       indexed_attribute_num < description.ExtensionSize(SMAIndexSubBlockDescription::indexed_attribute_id);
+       ++indexed_attribute_num) {
+    attribute_id indexed_attribute_id = description.GetExtension(
+        SMAIndexSubBlockDescription::indexed_attribute_id, indexed_attribute_num);
+    if (!relation.hasAttributeWithId(indexed_attribute_id)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+std::size_t SMAIndexSubBlock::EstimateBytesPerTuple(
+    const CatalogRelationSchema &relation,
+    const IndexSubBlockDescription &description) {
+  // TODO(marc): Returning 1 almost always gives too much space. An enhancement
+  // would be to inform the storage subblock that this index uses space w.r.t.
+  // # of attributes being indexed, not tuples.
+  return 1;
+}
+
+bool SMAIndexSubBlock::bulkAddEntries(const TupleIdSequence &tuples) {
+  header_->index_consistent = false;
+  return true;  // There will always be space for the entry.
+}
+
+void SMAIndexSubBlock::bulkRemoveEntries(const TupleIdSequence &tuples) {
+  header_->index_consistent = false;
+}
+
+bool SMAIndexSubBlock::addEntry(const tuple_id tuple) {
+  header_->index_consistent = false;
+  return true;  // There will always be space to insert the entry.
+}
+
+void SMAIndexSubBlock::removeEntry(const tuple_id tuple) {
+  header_->index_consistent = false;
+}
+
+bool SMAIndexSubBlock::rebuild() {
+  resetEntries();
+  header_->count_aggregate = 0;
+  if (tuple_store_.isPacked()) {
+    for (tuple_id tid = 0; tid <= tuple_store_.getMaxTupleID(); ++tid) {
+      addTuple(tid);
+    }
+  } else {
+    for (tuple_id tid = 0; tid <= tuple_store_.getMaxTupleID(); ++tid) {
+      if (tuple_store_.hasTupleWithID(tid)) {
+        addTuple(tid);
+      }
+    }
+  }
+  header_->index_consistent = true;
+  return true;
+}
+
+void SMAIndexSubBlock::addTuple(tuple_id tuple) {
```

Comment:
Remove can incur the cost of a scan, so we prefer reusing the rebuild() functionality.